xds/src/main/java/com/linecorp/armeria/xds/RouteResourceNode.java

```diff
@@ -58,27 +60,49 @@ public void doOnChanged(RouteXdsResource resource) {
         pending.clear();
         final RouteConfiguration routeConfiguration = resource.resource();
         int index = 0;
+        boolean added = false;
         for (VirtualHost virtualHost: routeConfiguration.getVirtualHostsList()) {
             for (Route route: virtualHost.getRoutesList()) {
                 if (route.getActionCase() != ActionCase.ROUTE) {
                     continue;
                 }
                 final RouteAction routeAction = route.getRoute();
-                final String cluster = routeAction.getCluster();
+                final String clusterName = routeAction.getCluster();
 
                 // add a dummy element to the index list so that we can call List.set later
                 // without incurring an IndexOutOfBoundException when a snapshot is updated
                 clusterSnapshotList.add(null);
                 pending.add(index);
-                final ClusterResourceNode node =
-                        new ClusterResourceNode(null, cluster, xdsBootstrap(),
-                                                resource, snapshotWatcher, virtualHost, route,
-                                                index++, ResourceNodeType.DYNAMIC);
-                children().add(node);
-                xdsBootstrap().subscribe(node);
+
+                final BootstrapClusters bootstrapClusters = xdsBootstrap().bootstrapClusters();
+                final ClusterSnapshot clusterSnapshot = bootstrapClusters.clusterSnapshot(clusterName);
+                if (clusterSnapshot != null) {
+                    final EndpointSnapshot endpointSnapshot = clusterSnapshot.endpointSnapshot();
```

Comment:
Question) Does it not work out of the box if just `staticCluster` for both `edsCluster` and `clusterSnapshot`?

Reply:
It's only possible if we fetch the load assignment from the control plane before it's used and block until the fetch is done.

Reply:
I see, I actually consider this a bug in the original implementation if this doesn't work 😅 
Let me try debugging this later 🙏

Reply:
Since `edsClusters` do not include endpoints (`ClusterLoadAssignment`), they do not have a `ClusterSnapshot` unless the bootstrap prefetches the corresponding EDS. Have I overlooked something? Or is the name `edsCluster` causing confusion? (By the way, I've renamed it to `edsConfigCluster`.)

Reply:
> they do not have a ClusterSnapshot unless the bootstrap prefetches the corresponding EDS.

I imagined that this would follow the same flow as normal EDS - so once the `ClusterResourceNode` sees a eds config, it would use the `ConfigSourceClient` to recursively fetch the endpoint information.
Even in the normal flow, the EDS information isn't prefetched and fetched lazily.

I didn't look carefully through this PR to check if there is any difference though

Reply:
Just checked out the branch with the following diff and the test suite seems to pass at the very least:
https://ge.armeria.dev/s/fs442gvoejus2

```java
ecorp/armeria/xds/BootstrapClusters.java
index 1cd003f92..193eadfa6 100644
--- a/xds/src/main/java/com/linecorp/armeria/xds/BootstrapClusters.java
+++ b/xds/src/main/java/com/linecorp/armeria/xds/BootstrapClusters.java
@@ -38,12 +38,8 @@ final class BootstrapClusters implements SnapshotWatcher<ClusterSnapshot> {
                 if (cluster.hasLoadAssignment()) {
                     // no need to clean this cluster up since it is fully static
                     StaticResourceUtils.staticCluster(xdsBootstrap, cluster.getName(), this, cluster);
-                } else if (cluster.hasEdsClusterConfig()) {
-                    edsConfigClusters.put(cluster.getName(), cluster);
-                } else {
-                    throw new IllegalArgumentException(
-                            "Cluster must has a load assignment or EDS cluster config.");
                 }
+                edsConfigClusters.put(cluster.getName(), cluster);
             }
         }
     }
diff --git a/xds/src/main/java/com/linecorp/armeria/xds/RouteResourceNode.java b/xds/src/main/java/com/linecorp/armeria/xds/RouteResourceNode.java
index 91a36cac4..d06496628 100644
--- a/xds/src/main/java/com/linecorp/armeria/xds/RouteResourceNode.java
+++ b/xds/src/main/java/com/linecorp/armeria/xds/RouteResourceNode.java
@@ -75,17 +75,6 @@ final class RouteResourceNode extends AbstractResourceNodeWithPrimer<RouteXdsRes
                 pending.add(index);
 
                 final BootstrapClusters bootstrapClusters = xdsBootstrap().bootstrapClusters();
-                final ClusterSnapshot clusterSnapshot = bootstrapClusters.clusterSnapshot(clusterName);
-                if (clusterSnapshot != null) {
-                    final EndpointSnapshot endpointSnapshot = clusterSnapshot.endpointSnapshot();
-                    assert endpointSnapshot != null;
-                    snapshotWatcher.snapshotUpdated(new ClusterSnapshot(
-                            clusterSnapshot.xdsResource(), endpointSnapshot,
-                            virtualHost, route, index++));
-                    added = true;
-                    continue;
-                }
-
                 final Cluster cluster = bootstrapClusters.edsConfigCluster(clusterName);
                 final ClusterResourceNode node;
                 if (cluster != null) {
user@AL02437565 upstream-armeria % 
```

Let me know if I'm misunderstanding anything

Reply:
>> they do not have a ClusterSnapshot unless the bootstrap prefetches the corresponding EDS.

> I imagined that this would follow the same flow as normal EDS - so once the ClusterResourceNode sees a eds config, it would use the ConfigSourceClient to recursively fetch the endpoint information.

Yes, it does follow the same flow if the cluster has the EDS config:
https://github.com/line/armeria/pull/5468/files#diff-90a45b0d8c8062bb063c60c0e8c5647c42222c7d3973dfe1f5a3a65a006657dcR92

`staticCluster` that is configured to the `Bootstrap` has either `ClusterLoadAssigment` or EDS cluster config.
If the cluster has `ClusterLoadAssigment`, the `ClusterSnapshot` of the cluster will be created right away with `EndpointSnapshot`. It doesn't have to fetch the EDS resource from the control plane.
However, if it doesn't, it can't create the `ClusterSnapshot` until its EDS information is fetched from the control plane. And currently, it's fetched when the cluster name is specified via `ClusterRoot` or by the `RouteResourceNode`.

Because of the difference, we have to use a different approach. Please correct me if I'm wrong. 😉

Reply:
Ah now I get it. :) I thought you were saying that it can fetch the `ClusterSnapshot` even for an EDS config cluster. That was my bad. 😓  
Yeah, I intentionally didn't do that because I didn't want to create a `ClusterResoureNode` but let me try that since it simplifies the code.